crypto/evp/p_lib.c

```diff
@@ -983,12 +983,19 @@ int EVP_PKEY_type(int type)
 
 int EVP_PKEY_get_id(const EVP_PKEY *pkey)
 {
+    if (pkey->type == EVP_PKEY_KEYMGMT) {
+        const char *name = EVP_KEYMGMT_get0_name(pkey->keymgmt);
+        int type = evp_pkey_name2type(name);
+        if (type != NID_undef)
+            return type;
+    }
+
     return pkey->type;
 }
 
 int EVP_PKEY_get_base_id(const EVP_PKEY *pkey)
 {
-    return EVP_PKEY_type(pkey->type);
+    return EVP_PKEY_type(EVP_PKEY_get_id(pkey));
```

Comment:
Why is this change needed?

Reply:
The base_id should be obtained with the correct id, if not, it will also encounter the situation where the type is EVP_PKEY_KEYMGMT.